doc(changelog): support to generate changelog through travis

mynktl opened the pull request:
Changes:
- Added script to trigger changelog PR from travis

```bash
Usage:
	./script/create_changelog <options> TAG-NAME

Example:
	./script/create_changelog v1.1.0-RC1

TAG-NAME is required to execute this script.

Supported options are as below. Mandatory arguments to long options are mandatory for short options too:
  -h, --help				Show this message and exit
  -r, --repo REPO			Github repo, in org/repo format.
					To raise a PR, this option is mandatory.
					If mentioned PR will be created in this repo.
					Example:
					  -r openebs/plugin
					  --repo openebs/plugin
  -p, --production-branch BRANCH  	name of production branch.
					Option -r|--repo is required to use this option.
					PR for changelog will be created against this mentioned branch
					If options -p|--production-branch and -c|--release-branch, both are
					mentioned then two PR will be raised.
					Example:
					  -p master
					  --production-branch master
  -c, --release-branch RELEASEBRANCH	name of release branch.
					Option -r|--repo is required to use this option.
					PR for changelog will be created against this mentioned branch
					If options -p|--production-branch and -c|--release-branch, both are
					mentioned then two PR will be raised.
					Example:
					  -c v1.10.x
					  --release-branch v1.10.x


This script expects following environment variable
- USERNAME
   This is your github username, configured through git config --global user.name
   This script doesn't print your USERNAME on output

- TOKEN
   This is github token. This script doesn't print your TOKEN on output.
   To create a new token,
   refer https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line

- EMAIL
   This is your email id, configured through git config --global user.email, to sign the commit
   This script doesn't print your EMAIL on output

How this script works?
- Generating changelog
   Script will use files from directory changelogs/unreleased to create a changelog
   It will create a new directory, named tag-name, under directory changelogs/released
   and move all the files from directory changelogs/unreleased to new directory.
   Once all the files are moved to new directory, it will create a temporary file
   in /tmp/ directory and copy the content of all files, under new directory to a
   temporary file.
   If tag is release tag, like v1.1.0, and there are existing directory for rc release,
   like v1.1.0-RC1, then script will copy the contents of all files, under this rc release
   directory, to a temporary file.
   After this, script will add the content of temporary file to root changelog file
   and delete the temporary file

- Creating a branch
  Script creates a new branch named changelog-timestamp, like changelog-20200512164456

- Creating a commit
   Script will create commit with following files using your USERNAME and EMAIL
   - repo-name/CHANGELOG.md (modified)
   - repo-name/changelogs/unreleased/* (deleted)
   - repo-name/changelogs/released/tag-name (new directory)

- Pushing branch to repo
   To push the changes to github, script will add new remote, named upstream,
   using your USERNAME and TOKEN

- Creating a pull request
   Script uses github API to create a pull request, using your USERNAME and TOKEN.
   Refer https://developer.github.com/v3/pulls/#create-a-pull-request for API
```

mynktl commented:
Sample PR: https://github.com/openebs/velero-plugin/pull/89

akhilerm commented on script/create_changelog:
```diff
@@ -0,0 +1,250 @@
+CHANGELOG_DIR=changelogs
+UNRELEASED_CHANGELOG_DIR=changelogs/unreleased
+RELEASED_CHANGELOG_DIR=changelogs/released
+ROOT_CHANGELOG=CHANGELOG.md
```
question: @mynktl  This [release file](https://github.com/openebs/velero-plugin/blob/master/RELEASE.md) mentions about a `velero-plugin/CHANGELOG-{TAG}.md` file . Are we actually using it?

mynktl replied:
No. I need opinion from other repo maintainers on that. Since we are following monthly release and two RC build, this will create lot of files on root directory within a year. Each month we will be adding 3 new files to root folder.

akhilerm commented on script/create_changelog:
```diff
@@ -0,0 +1,250 @@
+	git config --global user.email ${EMAIL}
+	local repo_name=$(echo $GIT_REPO | cut -d '/' -f2)
+	git remote add upstream https://${USERNAME}:${TOKEN}@github.com/$USERNAME/$repo_name.git > /dev/null 2>&1
+	git branch $BRANCH && git checkout $BRANCH
```
If we are on the release branch, suppose `v1.10.x`, then `git branch` will create a new branch from `v1.10.x` right? not master.

mynktl replied:
yes.

mynktl replied:
cherry_pick will add commit to production branch and delete old branch an create a new branch=$BRANCH from production branch
This is required because release branch and production branch may have different commits. Ideally master branch will have all the changelog, but release branch will have subset of that.

mynktl replied:
@akhilerm I've updated the script to raise a PR against release branch also.

kmova commented on script/create_changelog:
```diff
@@ -0,0 +1,250 @@
+	exit 1
+fi
+
+BRANCH=changelog-$(date +'%Y%m%d%H%M%S')
```
@mynktl -- would it be possible to move the change-log generate and rasing PR into different scripts. 

thinking of the following usecase:
- we have not yet fully mastered the best practicies around - writting good comments or organizing the fixes. this may take one or two more releases. 
- have an option for the repo maintainer to locally run the script, check the git diff, make the necessary modifications and decide to raise the PR.

kmova replied:
One practice - probably we need to adopt some of the comments addressed in this PR about changelogs appearing in master as well as release branch could be: 
- decide that it is ok for a PR to be present in both changelogs of current release - and next release
- don't add a changelog to master, if it is going to be cherry-picked (maybe change the practice to first fix in the release branch and then cherry-pick to master)

mynktl replied:
ok. I will fragment this script into two script -- one to create changelog branch and create commit and second to raise a PR.
> decide that it is ok for a PR to be present in both changelogs of current release - and next release

changelog will appear on master branch only. If someone wants to read changelog, He/She will open repo page only and that will redirect it to default branch.
> don't add a changelog to master, if it is going to be cherry-picked (maybe change the practice to first fix in the release branch and then cherry-pick to master)

This is a very rare scenario since. In this scenario owner can update the branch itself, used in PR.
> we have not yet fully mastered the best practicies around - writting good comments or organizing the fixes. this may 
take one or two more releases.

Agree. Writing good comments, I've added comments in script where i thought necessary. But it may depends on other users according to their knowledge of bash. It may vary. If i've not added comment on any complex logic which is difficult to understand, please drop a comment.
> have an option for the repo maintainer to locally run the script, check the git diff, make the necessary modifications and decide to raise the PR.

repo maintainer can run this script locally and update the PR. Fragmenting this script into two script -- to create commit and to raise a PR -- will solve this.

kmova replied:
can it be split as: 
- make changes 
- commit and raise PR. 

Once the script commits, it includes modifying and adding a new commit or squash related steps. 

re: Comments, I meant overall contributions to the repo, not just on this PR. They may be coming from new contributors at times, who have not yet fully understood the use-case or guidelines. Of course, even the most experienced folks cannot always get the comments right, as english doesn't have to be contributors' native language.

mynktl replied:
> -   make changes
> -  commit and raise PR.


I'm good with above approach. Thinking of adding some option in the script itself. Reason of this script is to automate the whole changelog process. If we split the script then maintainer need to do some extra work to execute second script.

mynktl replied:
About comments, PR review surely helps here, If comments are not placed(for complex code or critical section which requires additional measures for future changes) or not proper.

mynktl replied:
@kmova I've updated the script to generate changelog only and also if user want to raise a PR against release or production branch -- he/she can do that if arguments,with value, are mentioned.

mynktl commented:
> question: suppose after creating a tag (v1.10.0-RC1) from branch (v1.10.x), some changes are merged into master and cherry picked to v1.10.x branch. Now we have the same content in unreleased in 2 branches. The unreleased from `v1.10.x` will be moved to `v1.10.0-RC2` in the next tagging.
> 
> But if we don't remove the same content from the unreleased in master, won't it be added into the changelog for the next version (v1.11.0)

We are raising PR against the master. So it will remove the changelogs from unreleased folder. But i think we need to raise a cherry-pick for that PR to a release branch also, to remove the entries from unreleased.

kmova commented on script/create_changelog:
```diff
@@ -0,0 +1,402 @@
+#!/bin/bash
```
Add the copy right header.

kmova commented on script/create_changelog:
```diff
@@ -0,0 +1,402 @@
+		shift
+		GIT_REPO=$1
+		;;
+	-p|--production-branch)
```
n_: indent seems a little off here.

pawanpraka1 commented on script/create_changelog:
```diff
@@ -0,0 +1,402 @@
+	ls $UNRELEASED_CHANGELOG_DIR/* > /dev/null 2>&1 && cp $UNRELEASED_CHANGELOG_DIR/* $dest
+
+	cl=$(mktemp)
+	echo "$TAG / $(date +'%Y-%m-%d')" >> $cl
```
we are taking the date when we are creating the changelog, not when the tag was created.@kmova should it be date when tag was created or date when we are running the changelog?

kmova replied:
updating the changelog directly might result in merge conflicts in parallel PRs.

mynktl commented:
> @kmova @mynktl Few sugestions/comments. Developer has to add changelog in the unreleased folder, which we are going to parse and put it in the release folder as it is. If it makes sense, let the developers update the CHANGELOG.md file directly while raising the feature or bug fix PR, it should be must in each PR?

There are chances of conflict for CHANGELOG.md. We can't have it in CHANGELOG.md since it will be cosidered as released. We can have different file, but tracking it will require more efforts to maintain. 


> Also I see there has to be a PR with every release TAG, can we have one PR for main release which should have RC changelogs also?

Script will create only one PR for the given tag. If tag is stable one then it will include changelog from RC release also. But PR will be one.


> Also can we add the date when the tag was created, not when we are running the script? Can we get that info?

Good point! Script was meant to run in travis build so i've not added that functionality. But that is not the case now, I'll add the required changes.

mynktl commented:
Since this script is required for other repositories also, we planned to have this script under openebs/charts. I've raised a [PR](https://github.com/openebs/charts/pull/113) in openebs/charts to _gh-pages_ branch. 

@kmova @pawanpraka1 I've addressed your comments in [PR](https://github.com/openebs/charts/pull/113). PTAL.

Thanks.
